core/src/main/java/com/linecorp/armeria/common/stream/StreamMessageInputStream.java

```diff
@@ -0,0 +1,182 @@
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import static java.util.Objects.requireNonNull;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.function.Function;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.internal.common.stream.StreamMessageUtil;
+
+final class StreamMessageInputStream<T> extends InputStream {
+
+    public static <T> StreamMessageInputStream<T> of(
+            StreamMessage<T> source, Function<? super T, ? extends HttpData> httpDataConverter) {
+        return new StreamMessageInputStream<>(source, httpDataConverter);
+    }
+
+    private final StreamMessageInputStreamSubscriber<T> subscriber;
+    @Nullable
+    private InputStream inputStream;
+    private boolean closed;
+
+    private StreamMessageInputStream(StreamMessage<T> source,
+                                     Function<? super T, ? extends HttpData> httpDataConverter) {
+        requireNonNull(source, "source");
+        requireNonNull(httpDataConverter, "httpDataConverter");
+        subscriber = new StreamMessageInputStreamSubscriber<>(httpDataConverter);
+        source.subscribe(subscriber);
+    }
+
+    @Override
+    public int read() throws IOException {
+        return read(in -> {
+            try {
+                return in.read();
+            } catch (IOException e) {
+                return -1;
+            }
+        });
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        return read(in -> {
+            try {
+                return in.read(b, off, len);
+            } catch (IOException e) {
+                return -1;
+            }
+        });
+    }
+
+    private int read(Function<InputStream, Integer> function) throws IOException {
+        checkClosed();
+        if (inputStream == null || inputStream.available() == 0) {
+            try {
+                inputStream = subscriber.nextStream();
+            } catch (InterruptedException e) {
+                return -1;
+            }
+        }
+        return function.apply(inputStream);
+    }
+
+    @Override
+    public void close() throws IOException {
+        checkClosed();
```

Comment:
Other `InputStream` implementations allow invoking `close` multiple times.
How about adhering to that behavior?